### PR TITLE
Added support to log HTTP requests and responses.

### PR DESCRIPTION
**Pull request by prodis**
For some applications is valuable to have the option to log HTTP requests and responses. I have just added support to this.

What do you think?

### CODE REVIEW

**Comment by miry** [comment]
:+1:

**Comment by psychocandy** [comment]
looks good to me. @antlypls ?

File: lib/hasoffersv3/client.rb [hunk 1]
```diff
@@ -25,13 +25,16 @@ def request(http_method, target, method, params)
         http              = new_http(uri)
         raw_request       = Net::HTTP::Get.new(uri.request_uri)
       end
-      http_response = execute_request(http, raw_request)
+      http_response = execute_request(http, raw_request, uri)
```
Comment (antlypls) [inline lib/hasoffersv3/client.rb]: Not sure what's a point of modifying method passing extra arg, that is needed only for logging.
If you can surround this method call with logger.

Comment (prodis) [inline lib/hasoffersv3/client.rb]: @antlypls You are right. All params needed to log are in `Client#request`. Fixed.

File: lib/hasoffersv3/logger.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        s.string
+      end
+    end
+
+    def append_headers(message, headers)
+      headers.each do |name, value|
+        message.puts("#{name}: #{value}")
+      end
+
+      message.puts('')
```
Comment (antlypls) [inline lib/hasoffersv3/logger.rb]: I think just `message.puts` is enough

Comment (prodis) [inline lib/hasoffersv3/logger.rb]: It's true. Done.

File: lib/hasoffersv3/client.rb [hunk 3]
```diff
@@ -16,22 +16,27 @@ def request(http_method, target, method, params)
       target_uri = "#{base_uri}/#{target}.json"
 
       if http_method == :post
-        uri               = URI.parse(target_uri)
```
Comment (antlypls) [inline lib/hasoffersv3/client.rb]: These changes are just formatting, and not related to implemented feature.
So in general styling fixes should be in separate PR or at least in a separate commits.
Otherwise it messes up blame logs.

File: lib/hasoffersv3/client.rb [hunk 4]
```diff
@@ -16,22 +16,27 @@ def request(http_method, target, method, params)
...
+        uri              = URI.parse("#{target_uri}?#{query_string(data)}")
+        http             = new_http(uri)
+        raw_request      = Net::HTTP::Get.new(uri.request_uri)
       end
+
+      logger.log_request(raw_request, uri)
+
       http_response = execute_request(http, raw_request)
 
-      Response.new(http_response, @configuration.json_driver)
```
Comment (antlypls) [inline lib/hasoffersv3/client.rb]: same here, removing `@` is not necessary for implemented feature.

File: lib/hasoffersv3/client.rb [hunk 5]
```diff
@@ -16,22 +16,27 @@ def request(http_method, target, method, params)
...
+      logger.log_response(http_response)
+
+      Response.new(http_response, configuration.json_driver)
     end
 
     def execute_request(net_http, raw_request)
-      net_http.request raw_request
```
Comment (antlypls) [inline lib/hasoffersv3/client.rb]: Just style too.

File: lib/hasoffersv3/configuration.rb [hunk 6]
```diff
@@ -34,7 +35,7 @@ def self.default_json_driver
 
     attr_reader :options
 
-    def initialize(options={})
```
Comment (antlypls) [inline lib/hasoffersv3/configuration.rb]: Just style too.

File: lib/hasoffersv3/testing.rb [hunk 7]
```diff
@@ -43,7 +43,7 @@ def execute_request(net_http, raw_request)
       if HasOffersV3::Testing.enabled?
         HasOffersV3::Testing.execute_stubbed_request
       else
-        original_execute_request net_http, raw_request
```
Comment (antlypls) [inline lib/hasoffersv3/testing.rb]: Just style too.

**Comment by antlypls** [comment]
@prodis In general looks ok. But I suggest revert all style changes, and keep changes that are needed for logging feature. Squash all changes into 1 commit. After that we can merge it.
And if you want you can open separate PR with style fixes.

**Comment by prodis** [comment]
@antlypls I got it. Done.

**Comment by antlypls** [comment]
@prodis nice!
@psychocandy 👍

File: spec/lib/configuration_spec.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
+require 'spec_helper'
```
Comment (miry) [inline spec/lib/configuration_spec.rb]: @prodis can you remove this line. to be consistent with other tests.

Comment (prodis) [inline spec/lib/configuration_spec.rb]: @miry Done.

**Comment by miry** [comment]
@prodis Thank you
